chore: Update internal `tsconfig` presets and fix flagged type issues

kitten opened the pull request:
# Why

Related to https://github.com/expo/expo/pull/36019

See this link for the updated `tsconfig.preset.json` in `expo-module-scripts`: https://github.com/expo/expo/blob/a5e7c92b74f500140c341fe6ca9dd22055edc257/packages/expo-module-scripts/tsconfig.base.json

We'd like to verify an updated `tsconfig.json` configuration internally to prepare updating our TypeScript preset (for our templates) to enable stricter type checks and proper `moduleResolution: 'bundler'` mode.

To achieve this, we have to pass `strict: true` type checks, and use `moduleResolution: 'bundler'` ourselves to verify this works properly for ourselves first.

# How

- Enable `strict: true` and `moduleResolution: 'bundler'` (and some related `tsconfig` updates to the `expo-module-scripts` presets)
- Update all modules with type errors to pass type checks again
- Add `TODO` comments and notes where appropriate for things we should get back to

# Test Plan

- CI type checks

kitten committed:
Update tsconfig.json defaults to `moduleResolution: bundler`

kitten committed:
Add `strict: true` (and other adjustments) to base tsconfig.base.json

kitten committed:
eslint-config-universe: Turn off no-redeclare (and ts-variant of this rule) in TypeScript files

kitten committed:
expo-asset: Add missing module declaration for `@react-native/assets-registry/registry`

kitten committed:
expo-asset: Readd missing types for `@react-native/assets` (via `@types/react-native__assets`)

kitten committed:
expo-asset: Update for strict types and missing typings

kitten committed:
Replace `no-unused-*` eslint rules in favor of TypeScript config

kitten committed:
@expo/metro-runtime: Add missing module declarations

kitten committed:
@expo/metro-runtime: Fix typings and update for new strictness level

kitten committed:
@expo/metro-runtime: Rebuild

kitten committed:
expo-video: Add react-native-assets declaration reference and address null type error

kitten committed:
expo-audio: Add react-native-assets declaration reference and address unused error

kitten committed:
expo-auth-session: Fix type issues (CHANGELOG entry due to caught bug)

kitten committed:
expo-auth-session: Rebuild

kitten committed:
expo-av: Address type errors

kitten committed:
expo-battery: Add missing return type to events map

kitten committed:
expo-battery: Rebuild

kitten committed:
expo-blur: Fix style copy and webkit-specific property type

kitten committed:
expo-calendar: Add type utils and safety for stringifyDateValues

kitten committed:
expo-camera: Fix type errors due to strict mode

kitten committed:
expo-checkbox: Ignore missing react-native-web types

kitten committed:
expo-clipboard: Fix up error type and name access

kitten committed:
expo-constants: Update window key check

kitten committed:
expo-contacts: Fix typing of removeIds

kitten committed:
expo-dev-launcher: Add missing types

kitten committed:
expo-dev-menu: Add missing return type for event map type

kitten committed:
expo-device: Add missing @types/ua-parser-js dev dependency

kitten committed:
expo-file-system: Add return types to event map type

kitten committed:
expo-font: Add missing @types/fontfaceobserver and fix up constructor call

byCedric committed:
[@expo/html-elements]: Fixed missing types for React Native Web internals

kitten committed:
expo-gl: Work around type issues

byCedric committed:
expo-web-browser: add missing `buffer` type in `bufferToString`

byCedric committed:
expo-sqlite: add missing `error` and `row` types to utils

kitten committed:
expo-image: Resolve type errors

kitten committed:
exp-asset: Mark width,height as optional properties (related to expo-image)

kitten committed:
expo-image-manipulator: Add missing clamp argument types

kitten committed:
expo-image-picker: Add missing options type

kitten committed:
expo-linking: Add missing type annotations

kitten committed:
expo-localization: Add missing `@types/rtl-detect` package and fix up type errors

kitten committed:
expo-location: Fix type errors and missing types

kitten committed:
expo-mail-composer: Fix implicit type cast

kitten committed:
expo-maps: Leave todo note for implicit any type cast

kitten committed:
expo-media-library: Add missing type assertions

kitten committed:
expo-media-library: Leave todo comment

kitten committed:
expo-modules-core: Fix type issues

kitten committed:
expo-network: Add missing return type to events map type

kitten committed:
expo-notifications: Fix type errors and leave TODO notes

byCedric committed:
expo-speech: add missing types for speech events and expect errors for found issues

byCedric committed:
expo-sensors: add missing event handler types and fix `timeStamp` issue for `Gyroscope` on web

kitten committed:
expo-router: Fix type issues (TODOs left & important!)

kitten committed:
expo-router: Add missing `@types/semver`

kitten committed:
chore: Dedplicate `@types/semver`

byCedric committed:
expo-screen-orientation: add missing legacy screen handlers type exclusions

kitten committed:
expo: Update typings and type errors

kitten committed:
Apply lints

kitten committed:
expo-updates: Fix test types

kitten committed:
expo-standard-web-crypto: Fix mock type error

kitten committed:
Add missing error type in expo-sms

kitten committed:
expo-image-picker: Bandaid type error in ExponentImagePicker-test.web.ts

kitten committed:
expo-sqlite: Fix remaining type issues in tests

kitten committed:
expo-file-system: Fix type issue in FileSystem-test.native.ts

kitten committed:
expo-sensors: Add missing type annotations in Magnetometer-test.native.ts

kitten committed:
expo-modules-core: Fix resolve type in PermissionsHook-test.ts

kitten committed:
expo-asset: Resolve type issue (in test and src) and add missing @types/node dev dependency

kitten committed:
expo-screen-orientation: Fix type inference issue in ScreenOrientation-test.ios.ts

kitten committed:
expo-localization: Fix test type issues

kitten commented:
Update: A lot of packages that make us of Jest will now have a `src/ts-declarations/react-native.d.ts` file (or something similar) that references types that `expo` and `expo-module-scripts` use and declare for local use. This is always identical and is currently a **workaround**

Optimally, I'd like for our Jest preset not to use `ts-jest` and our tsconfigs not to exclude `__mocks__` and `__tests__`, however, refactoring this will require more time.

This will be needed because `tsc` is now unable to identify type issues, which also affects editors. Type issues won't be raised unless Jest runs these unit tests. Instead, Jest should be set up to run more similarly to our Metro pipeline (via `babel-jest`, the default otherwise)

Currently, jest and the `ts-jest` transform override means that type resolution and type checks are slightly different in Jest, and this workaround declaration is unfortunately needed to get this PR unblocked.

kitten committed:
eslint: Move unused overrides to expo-module-scripts

NOTE: We should make this the default eventually

kitten committed:
expo-blur: Apply lints

kitten committed:
expo-router: Apply lints

kitten committed:
expo-modules-core: Apply lints

kitten committed:
expo-image: Apply lints

kitten committed:
expo-file-system: Apply lints

kitten committed:
expo-clipboard: Apply lints

kitten committed:
expo-camera: Apply lints

kitten committed:
expo-av: Apply lints

kitten committed:
expo: Apply lints

kitten committed:
@expo/metro-runtime: Apply lints

kitten committed:
chore: Rebuild

- expo-asset
- expo-audio
- expo-auth-session
- expo-av
- expo-blur
- expo-calendar
- expo-camera
- expo-checkbox
- expo-clipboard
- expo-constants
- expo-dev-launcher
- expo-dev-menu
- expo-file-system
- expo-font
- expo-gl
- expo-image-manipulator
- expo-image-picker
- expo-linking
- expo-localization
- expo-location
- expo-mail-composer
- expo-maps
- expo-media-library
- expo-modules-core
- expo-network
- expo-notifications
- expo-router
- expo-screen-orientation
- expo-sensors
- expo-speech
- expo-sqlite
- expo-video
- expo-web-browser
- expo
- @expo/html-elements

kitten committed:
Dedup/update yarn.lock

kitten committed:
expo-asset: Fix test type error

kitten committed:
expo-updates/e2e: Test adding tsconfig.json to expo-updates/e2e/

kitten committed:
expo-location: Update test type issues

kitten committed:
expo: Fix type issues in tests

kitten committed:
Add localized d.ts references

kitten committed:
expo-font: Add typing for `@react-native/assets-registry/registry`

kitten committed:
expo-constants: Fix type test errors in Constants-test.ts

kitten committed:
expo: Patch RequestUtils-test.ts

kitten committed:
expo-font: Update asset entry mock for tests

kitten committed:
expo: Rebuild

byCedric commented on packages/expo-sqlite/src/__tests__/hooks-test.ios.tsx:
```diff
@@ -9,7 +9,7 @@ jest.mock('../ExpoSQLite', () => require('../__mocks__/ExpoSQLite'));
 describe(useSQLiteContext, () => {
   it('should return a SQLite database instance', async () => {
-    const wrapper = ({ children }) => (
+    const wrapper = ({ children }: { children: React.ReactNode }) => (
```
I highly prefer `({ children }: PropsWithChildren)` over typing `{ children: React.ReactNode }` - technically should be `{ children?: React.ReactNode }`.

kitten replied:
no strong preference here, but fair enough. I've been burned by over-use of `@types/react`-provided types before, and usually now stick with built-in JSX namespace types, with the exception of `React.ReactNode` (or React-specific ones as needed). This also makes a lot of the types as specific as they have to be

byCedric reviewed (approved):
Left a few nits, but nothing merge-blocking.

kitten commented on packages/expo-av/src/FullscreenUtils.web.ts:
```diff
@@ -1,3 +1,12 @@
+declare global {
+  interface Document {
```
there's a small difference. Using a namespace locally replaces the type and otherwise works like a TS namespace. This keeps the variable isolated.

If you wanted to declare a global variable that's then available on `globalThis` you have to exactly write `declare global { var ... }` (var specifically)
However, for tests this isn't always great even if we wanted to do this since it can leak out of the file and affect other modules

kitten commented on packages/expo-screen-orientation/src/ExpoScreenOrientation.web.ts:
```diff
@@ -143,27 +155,35 @@ class ExpoScreenOrientation extends NativeModule<ExpoOrientationEvents> {
+
+    // See: https://developer.mozilla.org/en-US/docs/Web/API/Screen/unlockOrientation
+    const _legacyUnlockUniversal: undefined | (() => boolean) =
+      // @ts-expect-error - These legacy APIs are removed from the types
+      screen.unlockOrientation || screen.mozUnlockOrientation || screen.msUnlockOrientation;
```
I couldn't be arsed to be honest 😅 There's a lot of lot todos left for these mainly because this kind of code is repeated and copied often and the difference between typing it properly or ignoring it was small.

I can align it though if we really want it and don't think we want to remove these for some legacy browsers

kitten commented on packages/expo-checkbox/src/ExpoCheckbox.web.tsx:
```diff
@@ -10,6 +10,7 @@
  */
 
 import React from 'react';
+// @ts-expect-error: TODO(@kitten): We optimally shouldn't pull in all of react-native-web just for expo-checkbox. It's excessive
 import { StyleSheet, View, unstable_createElement as createElement } from 'react-native-web';
```
Optimally, I'd like to avoid sharing them as much as possible to enable future refactors to be smoother and to avoid types leaking at a distance.

I don't mind copying the file over but my thinking here is:
- I'd like to avoid copying the whole declaration
- I'd like to avoid relying on a custom type root
- I don't want to necessarily add a reference across packages that aren't related

I think I also ran into type issues here and `expo-checkbox` is a bit problematic so I moved on

kitten replied:
^Resolving; Let's leave it at the ignore for now. I'm pretty sure I'm running into more issues here even if it's retyped

kitten commented on packages/expo-asset/src/resolveAssetSource.ts:
```diff
@@ -42,4 +49,6 @@ Object.defineProperty(resolveAssetSource, 'setCustomSourceTransformer', {
   },
 });
 
+export default resolveAssetSource as resolveAssetSource;
```
The interface overlay didn't work and I wanted to avoid changing the implementation in this exact case. Ultimately, I don't really care too much how it's written as long as the correct type is exported and as little as possible is casted via any or ignored for APIs that are exported

kitten committed:
expo-screen-orientation: Narrow polyfill types for legacy browsers

kitten committed:
expo-screen-orientation: Rebuild

kitten committed:
expo-asset: Fix up import path with redirected types

kitten committed:
expo-asset: Rebuild

kitten commented on packages/expo-asset/src/ts-declarations/react-native-assets.d.ts:
```diff
@@ -41,3 +41,17 @@ declare module 'react-native/Libraries/Image/resolveAssetSource' {
 }
+
+declare module '@react-native/assets-registry/registry' {
+  import type { PackagerAsset } from '@react-native/assets/registry';
```
Provided in typings as per DMs. The `definitely-typed` package is wrong here about the name, but it's still advantageous to use it for now

kitten commented on packages/expo-file-system/src/__tests__/FileSystem-test.native.ts:
```diff
@@ -33,7 +33,7 @@ describe('FileSystem', () => {
         remoteUri,
         localUri,
-        downloadResumable._uuid,
+        (downloadResumable as any)._uuid,
```
We do but I think it was marked as private, so imho, in tests, this is acceptable, given that it's doing an equality check here in jest's matchers

kitten commented on packages/expo-dev-launcher/src/setUpErrorHandler.fx.ts:
```diff
@@ -1,5 +1,3 @@
-import { ErrorUtils } from 'react-native';
```
Just checked. The typings mark this as type-only and the typings happen to define it globally, I think? Weird situation. I'll patch this with a typings file

kitten replied:
Ah no, ok, so this is indeed just a type, and the export is unnamed, weirdly enough. It might of course exist, given that it worked before, but react-native's typings here are incomplete. However, it seems this also exists as a global, so this is why TS suggested to remove this here

kitten committed:
expo-sqlite: Update props types

lukmccall reviewed (approved):
Packages that I’m familiar with look good.

kitten committed:
expo-asset: Unsafely cast type and leave todo note

cc @Kudo

kitten committed:
expo-web-browser: Remove generic on Uint8Array

We should avoid this TS feature. The generic here has changed, and
serves little to no purpose. `SharedArrayBuffer` is unsound and
broken.

kitten commented:
CI run on this commit is passing and green: https://github.com/expo/expo/pull/36075/commits/87fd555e42bd7fa36a034f5c2f154e0b541758f0

The commit after is a "hotfix" for a type issue that affects the docs, but waiting for CI for a small change is too costly and takes too long to delay this further
